### PR TITLE
palettized workshop 1

### PR DESCRIPTION
**Pull request by PatrikLundell**
#### Summary

None

#### Purpose of change

Make use of palette parameters to shrink JSON footprint.

Bycatches:
- Added concrete as a construction material.
- Fixed buggy cross reference to the livestock expansion (copy-paste(?) error) plus addition of unused blueprint.
- Removed now redundant (and conflicting) entries from Aftershock.

#### Describe the solution

Same as for other base camps/expansions, i.e. use palette parameters to reference palettes instead of duplicating each blueprint with different palettes.
Cleaned up palettes to only contain what's actually used.

#### Describe alternatives you've considered

Decouple construction from furniture, but considered that to be out of scope.

#### Testing

Created the complete expansion made out of concrete in the NW expansion slot. Inspected the results and killed the game. Repeated the process with a new material in a different expansion slot (slot change isn't necessary as there is no location dependence in the blueprints, but the work had to be done somewhere) constructing only what's necessary to complete all buildings. Inspection and killing of the game again. Repeated until all materials had been covered.

The Aftershock changes were tested by generating an Aftershock world and stumble through to the surface, setup a base camp and a workshop 1 expansion. Constructed the first (mi-go resin) section successfully, built the first set of stuff inside. Verified that the anvil construction recipe indeed required two anvils and removed the override for that recipe as well, verifying that it now required only one anvil. I stopped the testing there, as it should be identical to the non Aftershock one from there on.

#### Additional context

It can be noted that you can't make use of dual palettes to deal with the furniture as the construction also specifies the floor underneath. Since some of the elements that don't differ between material palettes are overloaded by the common blueprint but without a floor you would have a collision.

Note that I haven't looked at whether the remaining entries in the Aftershock file have become redundant or not. That may be something the Aftershock maintainers may want to look at.

### CODE REVIEW

**Comment by Maleclypse** [comment]
```(all_mods)=> (continued from above) ERROR : src/mapgen.cpp:524 [const update_mapgen &string_id<update_mapgen>::obj() const [T = update_mapgen]] Using invalid nested_mapgen_id fbmw_room6_rock_north
(all_mods)=> 10:13:41.878 ERROR : src/recipe.cpp:473 [cata::value_ptr<parameterized_build_reqs> calculate_all_blueprint_reqs(const update_mapgen_id &, const std::map<std::string, std::map<std::string, translation> > &)] Requested argument 1 but input has only 1 at: "update_mapgen %s used for blueprint, but has %zu| versions, where it should have exactly one"
(all_mods)=> 10:13:41.878 ERROR : src/mapgen.cpp:524 [const update_mapgen &string_id<update_mapgen>::obj() const [T = update_mapgen]] Using invalid nested_mapgen_id fbmw_room6_log_north
(all_mods)=> 10:13:41.878 ERROR : src/recipe.cpp:473 [cata::value_ptr<parameterized_build_reqs> calculate_all_blueprint_reqs(const update_mapgen_id &, const std::map<std::string, std::map<std::string, translation> > &)] Requested argument 1 but input has only 1 at: "update_mapgen %s used for blueprint, but has %zu| versions, where it should have exactly one"
(all_mods)=> 10:13:41.878 ERROR : src/mapgen.cpp:524 [const update_mapgen &string_id<update_mapgen>::obj() const [T = update_mapgen]] Using invalid nested_mapgen_id fbmw_room6_migo_resin_north
(all_mods)=> 10:13:41.878 ERROR : src/recipe.cpp:473 [cata::value_ptr<parameterized_build_reqs> calculate_all_blueprint_reqs(const update_mapgen_id &, const std::map<std::string, std::map<std::string, translation> > &)] Requested argument 1 but input has only 1 at: "update_mapgen %s used for blueprint, but has %zu| versions, where it should have exactly one"
(all_mods)=> 10:13:41.878 ERROR : src/mapgen.cpp:524 [const update_mapgen &string_id<update_mapgen>::obj() const [T = update_mapgen]] Using invalid nested_mapgen_id fbmw_room6_rammed_earth_north
(all_mods)=> 10:13:41.878 ERROR : src/recipe.cpp:473 [cata::value_ptr<parameterized_build_reqs> calculate_all_blueprint_reqs(const update_mapgen_id &, const std::map<std::string, std::map<std::string, translation> > &)] Requested argument 1 but input has only 1 at: "update_mapgen %s used for blueprint, but has %zu| versions, where it should have exactly one"
(all_mods)=> 10:13:41.878 ERROR : src/mapgen.cpp:524 [const update_mapgen &string_id<update_mapgen>::obj() const [T = update_mapgen]] Using invalid nested_mapgen_id fbmw_room6_wad_north
(all_mods)=> 10:13:41.878 ERROR : src/recipe.cpp:473 [cata::value_ptr<parameterized_build_reqs> calculate_all_blueprint_reqs(const update_mapgen_id &, const std::map<std::string, std::map<std::string, translation> > &)] Requested argument 1 but input has only 1 at: "update_mapgen %s used for blueprint, but has %zu| versions, where it should have exactly one"
(all_mods)=> 10:13:41.878 ERROR : src/mapgen.cpp:524 [const update_mapgen &string_id<update_mapgen>::obj() const [T = update_mapgen]] Using invalid nested_mapgen_id fbmw_room6_wood_north
(all_mods)=> 10:13:41.878 ERROR : src/recipe.cpp:473 [cata::value_ptr<parameterized_build_reqs> calculate_all_blueprint_reqs(const update_mapgen_id &, const std::map<std::string, std::map<std::string, translation> > &)] Requested argument 1 but input has only 1 at: "update_mapgen %s used for blueprint, but has %zu| versions, where it should have exactly one"
(all_mods)=> 10:13:46.125 ERROR : src/recipe_dictionary.cpp:675 [static void recipe_dictionary::check_consistency()] recipe faction_base_modular_workshop_room6_rock_north specifies invalid construction_blueprint fbmw_room6_rock_north; that should be a defined update_mapgen_id but is not
(all_mods)=> 10:13:46.126 ERROR : src/recipe_dictionary.cpp:675 [static void recipe_dictionary::check_consistency()] recipe faction_base_modular_workshop_room6_log_north specifies invalid construction_blueprint fbmw_room6_log_north; that should be a defined update_mapgen_id but is not
(all_mods)=> 10:13:46.126 ERROR : src/recipe_dictionary.cpp:675 [static void recipe_dictionary::check_consistency()] recipe faction_base_modular_workshop_room6_migo_resin_north specifies invalid construction_blueprint fbmw_room6_migo_resin_north; that should be a defined update_mapgen_id but is not
(all_mods)=> 10:13:46.126 ERROR : src/recipe_dictionary.cpp:675 [static void recipe_dictionary::check_consistency()] recipe faction_base_modular_workshop_room6_rammed_earth_north specifies invalid construction_blueprint fbmw_room6_rammed_earth_north; that should be a defined update_mapgen_id but is not
(all_mods)=> 10:13:46.126 ERROR : src/recipe_dictionary.cpp:675 [static void recipe_dictionary::check_consistency()] recipe faction_base_modular_workshop_room6_wad_north specifies invalid construction_blueprint fbmw_room6_wad_north; that should be a defined update_mapgen_id but is not
(all_mods)=> 10:13:46.126 ERROR : src/recipe_dictionary.cpp:675 [static void recipe_dictionary::check_consistency()] recipe faction_base_modular_workshop_room6_wood_north specifies invalid construction_blueprint fbmw_room6_wood_north; that should be a defined update_mapgen_id but is not
```
This error seems like it might be related to your changes?

**Comment by PatrikLundell** [comment]
Edit: Hm, it seems I failed to actually post this comment...
In short, I don't understand the failures, as they refer to things removed.

I'm confused by the build failures of General build matrix / Basic Build and Test (Clang 10, Ubuntu, Curses) (pull_request). as it makes references to elements in files that were deleted by the PR (and I've been unable to find the first one elsewhere, in case of potential spaghetti referencing).
This suggests that there is an issue with testing using caches that aren't properly cleared/updated.

**Comment by PatrikLundell** [comment]
OK. Found where the removed reference was: somewhere in Aftershock. I'll have to see what to do about that. Since it's a mandatory test failing I don't think I need to put this in Draft while doing that.

**Comment by Maleclypse** [comment]
> OK. Found where the removed reference was: somewhere in Aftershock. I'll have to see what to do about that. Since it's a mandatory test failing I don't think I need to put this in Draft while doing that.
> 
> Edit: I think I need some help from whoever is managing Aftershock here:
> 
> As far as I can see, the only difference between ./data/mods/Aftershock/recipes/basecamps/recipe_overwrites.json:faction_base_modular_workshop_room6_rock_north and the original version in workshop 1 is that the workshop 1 version had a blueprint_name string that mentions the construction material (which I suspect may have been added after the Aftershock copy was created, possibly by me).
> 
> Thus, it seems to me that the Aftershock workshop material referencing overrides are currently completely redundant, and should be removed (the "common" one adds an anvil as a component, which makes a different, although it may be that this means two anvils are required due to later changes).

I think that's the correct action.  I'm one of the Aftershock managers though I need to do more for it soon.  You can make that change in this PR if you'd like.

**Comment by PatrikLundell** [comment]
I've updated the description to cover the Aftershock modifications as well.

**Comment by PatrikLundell** [comment]
The failures are due to someone accepting something into Master that shouldn't have been let in. I get the same errors after having downloaded the latest master (slightly before the time of this writing). I'd guess it's someone being too aggressive testing for redundant entries in JSON, alternatively adding the test before conducting the aggressive pruning that should have been included in the change.

**Comment by Maleclypse** [comment]
> The failures are due to someone accepting something into Master that shouldn't have been let in. I get the same errors after having downloaded the latest master (slightly before the time of this writing). I'd guess it's someone being too aggressive testing for redundant entries in JSON, alternatively adding the test before conducting the aggressive pruning that should have been included in the change.

It was test mismatch.  Both PRs were green independently because they ran against the same set of tests two different mergers merged the separate PRs.  Combined they made a problem that was now visible.  The fix has already been merged.

File: data/json/mapgen/basecamps/expansion/modular_workshop/version_1/modular_workshop_common.json [hunk 1]
```diff
@@ -34,9 +18,9 @@
   },
   {
     "type": "mapgen",
-    "update_mapgen_id": "faction_base_modular_workshop_0",
+    "update_mapgen_id": "faction_base_workshop_0",
```
Comment (Maleclypse) [inline data/json/mapgen/basecamps/expansion/modular_workshop/version_1/modular_workshop_common.json:21]:
```suggestion
    "update_mapgen_id": "faction_base_workshop_0"
```

File: data/json/mapgen/basecamps/expansion/modular_workshop/version_1/modular_workshop_common.json [hunk 2]
```diff
@@ -34,9 +18,9 @@
   },
   {
     "type": "mapgen",
-    "update_mapgen_id": "faction_base_modular_workshop_0",
+    "update_mapgen_id": "faction_base_workshop_0"
```
Comment (Maleclypse) [inline data/json/mapgen/basecamps/expansion/modular_workshop/version_1/modular_workshop_common.json:21]:
```suggestion
    "update_mapgen_id": "faction_base_workshop_0",
```

**Pull request merged**